### DISCUSSION TITLE
Who is responsibilty of all perpherals initialization?

### DISCUSSION THREAD

**Discussion by xiyanni** [polarfire-soc/polarfire-soc-documentation#118] [HSS and U-Boot] [answered]
When user puts the .xml file to the HSS project, does the HSS do initialization of all peripherals selected?
Such as DDR3L is checked in MSS Configurator. 
When we use the bare-metal example, such as : https://github.com/polarfire-soc/polarfire-soc-bare-metal-examples/tree/main/driver-examples/mss/mpfs-hal/mpfs-hal-ddr-demo
User update .xml to the demo, I called the it mpfs-hal-ddr-demo-new, Does this mpfs-hal-ddr-demo-new take place of HSS? And then, Is the DDR training contained in bare metal of E51 project?
How could make sure the bare metal example inculde all functions of HSS?

Customer has followed the step, after they programmed new bin file, board could not debug, "ELF file linked incorrectly" what's the reason?

**Comment by hughbreslin** [comment, marked as answer]
Hi @xiyanni 

I think there are 4 questions here:

1. Q: Does the HSS turn on / off any MSS peripherals that are enabled in the configurator 
A: no this is configured in the sys_setup.c but will be moved to be dynamic based on the XML in the future

2. Q: If you put updated XML in a bare metal project with new XML how does it translate to the HSS
A: if you're testing a project in debug and update the XML the changes will take effect, if you're using a project as a payload (i.e release) the updated XML needs to go in the HSS and the bare metal project will not have any effect on MSS settings / configuration. Have a look at the [build configurations](https://github.com/polarfire-soc/polarfire-soc-bare-metal-examples#build-configurations) for more info on this

3. Q: Who does DDR training
A: this is typically run by the E51 as its the first hart but could be run by any hart (in the HSS its the E51)

4. Q: Could not debug due to "ELF file linked incorrectly"
A I have no idea without looking at the project, what steps have been taken to debug this?

For more info on getting started we have a selection of [training videos](https://www.youtube.com/c/MicrochipDeveloperHelp/playlists?view=50&sort=dd&shelf_id=4) we made along with a [training series](https://www.microchip.com/en-us/education/technical-learning-center/webinars/fpga-webinars/risc-v-innovation-unleashed) we ran.
